addons/account/tests/test_account_payment_register.py

```diff
@@ -743,6 +744,27 @@ def test_register_payment_constraints(self):
                 .with_context(active_model='account.move', active_ids=self.out_invoice_2.ids)\
                 .create({})
 
+    def test_register_payment_multi_currency_conversion(self):
+        ''' When registering a payment using a different currency than the invoice, the amount should update when
+        changing wizard date
+        '''
+
+        PaymentRegister = self.env['account.payment.register']
+        data = [
+            (self.out_invoice_2, self.company_data['currency'], "invoice is in foreign currency and payment is in company currency"),
+            (self.out_invoice_3, self.currency_data['currency'], "invoice is in company currency and payment is in foreign currency"),
+            (self.out_invoice_2, self.currency_data_3['currency'], "invoice and payment are in different foreign currencies"),
+        ]
+        for invoice, pay_currency, msg in data:
+            inv_currency = invoice.currency_id
+            wizard = Form(PaymentRegister.with_context(active_model='account.move', active_ids=invoice.ids))
+            wizard.currency_id = pay_currency
+            for date in ('2017-01-01', '2016-01-01'):
+                payment_date = fields.Date.from_string(date)
+                wizard.payment_date = payment_date
+                expected_amount = inv_currency._convert(invoice.amount_residual, pay_currency, wizard.company_id, payment_date)
+                self.assertEqual(wizard.amount, expected_amount, f"Amount in payment wizard when {msg} is not correct")
+
```

Comment:
I think this part with the results we are waiting for should be hard-coded in the datalist.

Currently, the testing code contains part that are the same than the tested code.
If this part fails, then we wouldn't detect it.